src/code/z_moji.c

```diff
@@ -1,100 +1,114 @@
+ */
+
 #include "global.h"
 
+#define CHAR_WIDTH 8
+#define CHAR_HEIGHT 8
+
 u32 sFontColorRed = 255;
 u32 sFontColorGreen = 255;
 u32 sFontColorBlue = 255;
 u32 sFontColorAlpha = 255;
-s32 D_80120120 = 0;
-s32 D_80120124 = 0;
 
-UNK_TYPE D_8015FFC0;
-UNK_TYPE D_8015FFC4;
+s32 sScreenPosX = 0;
+s32 sScreenPosY = 0;
+
+s32 sCurTLUTIndex;
+
+static s32 sUnused;
 
-void func_8007B910(u32 red, u32 green, u32 blue, u32 alpha) {
+void Moji_SetColor(u32 red, u32 green, u32 blue, u32 alpha) {
     sFontColorRed = red;
     sFontColorGreen = green;
     sFontColorBlue = blue;
     sFontColorAlpha = alpha;
 }
 
-void func_8007B934(s32 arg0, s32 arg1) {
-    if (arg0 > 39) {
-        D_80120120 = 39 * 8;
-    } else if (arg0 < 0) {
-        D_80120120 = 0;
+void Moji_SetPosition(s32 gridX, s32 gridY) {
+    if (gridX > (SCREEN_WIDTH / CHAR_WIDTH - 1)) {
+        sScreenPosX = SCREEN_WIDTH - CHAR_WIDTH;
+    } else if (gridX < 0) {
+        sScreenPosX = 0;
     } else {
-        D_80120120 = arg0 * 8;
+        sScreenPosX = gridX * CHAR_WIDTH;
     }
 
-    if (arg1 > 29) {
-        D_80120124 = 29 * 8;
-    } else if (arg1 < 0) {
-        D_80120124 = 0;
+    if (gridY > (SCREEN_HEIGHT / CHAR_HEIGHT - 1)) {
+        sScreenPosY = SCREEN_HEIGHT - CHAR_HEIGHT;
+    } else if (gridY < 0) {
+        sScreenPosY = 0;
     } else {
-        D_80120124 = arg1 * 8;
+        sScreenPosY = gridY * CHAR_HEIGHT;
     }
 }
 
-void func_8007B9A4(GraphicsContext* gfxCtx, u8 arg1) {
+void Moji_DrawChar(GraphicsContext* gfxCtx, char c) {
     s32 pad[2];
 
     OPEN_DISPS(gfxCtx, "../z_moji.c", 86);
 
-    if ((u32)gLetterTLUT & 0xF) {
-        osSyncPrintf("moji_tlut --> %X\n", gLetterTLUT);
+    if ((u32)gMojiFontTLUTs & 0xF) {
+        osSyncPrintf("moji_tlut --> %X\n", gMojiFontTLUTs);
     }
 
-    if (D_8015FFC0 != (arg1 & 3)) {
-        gDPLoadTLUT(POLY_OPA_DISP++, 16, 256, &gLetterTLUT[arg1 & 3]);
-        D_8015FFC0 = arg1 & 3;
+    if (sCurTLUTIndex != (c & 3)) {
+        gDPLoadTLUT(POLY_OPA_DISP++, 16, 256, &gMojiFontTLUTs[c & 3]);
+        sCurTLUTIndex = c & 3;
     }
 
-    gSPTextureRectangle(POLY_OPA_DISP++, D_80120120 << 2, D_80120124 << 2, (D_80120120 + 8) << 2, (D_80120124 + 8) << 2,
-                        G_TX_RENDERTILE, (u16)(arg1 & 4) * 64, (u16)(arg1 >> 3) * 256, 1 << 10, 1 << 10);
+    gSPTextureRectangle(POLY_OPA_DISP++, sScreenPosX << 2, sScreenPosY << 2, (sScreenPosX + 8) << 2, (sScreenPosY + 8) << 2,
+                        G_TX_RENDERTILE, (u16)(c & 4) * 64, (u16)(c >> 3) * 256, 1 << 10, 1 << 10);
 
     CLOSE_DISPS(gfxCtx, "../z_moji.c", 123);
 }
 
-void func_8007BBA8(GraphicsContext* gfxCtx, u8* arg1) {
+/**
+ * Does not work as is in most cases.
+ * Can work if the render mode, combiner and possibly other settings are set correctly.
+ * For example this works with the render mode used in `GfxPrint_InitDlist`, and `G_CC_MODULATEI_PRIM` for both combiner cycles.
+ */
+void Moji_DrawString(GraphicsContext* gfxCtx, char* str) {
```

Comment:
If this is a string, this should be a `const char*`, right?
```suggestion
void Moji_DrawString(GraphicsContext* gfxCtx, const char* str) {
```
Don't forget to update the prototype in functions.h too

Reply:
I have no clue if it should be `char*` or `const char*`, if you're sure or if someone else confirms I'll change it

These functions don't have a prototype in functions.h since they're all unused, less work for me FeelsOkayMan

Reply:
You could add them to functions.h, they may be useful to modders in the future